### PR TITLE
SSL Dependency update

### PR DESCRIPTION
**Pull request by g2vinay**
This dependency is expected to provide performance boost of 35-40% in RPS.
Running Storage Tests to verify and provide results.

### CODE REVIEW

**Comment by joshfree** [comment]
This is a great improvement and something that I would like to ship at some point very soon.  However we're hours away from locking down for the November SDK Update.  I think this change could use some bake time in master, and I would suggest that we merge this after November SDK Update is cut from master; so that it can bake and ship in the December SDK Update release.

/cc @Petermarcu as FYI

**Comment by anuchandy** [comment]
nice @g2vinay! it might be good to mentioned in readme( or equivalent) about this dependency that azure-core is pulling transitively for the user application, so that users concerned about size can exclude it and fallback to JDK ssl.

**Comment by anuchandy** [comment]
The tc-native seems a uber jar containing bits for linux/Mac/windows: https://repo1.maven.org/maven2/io/netty/netty-tcnative/2.0.25.Final/

We might want to think about applying `classifier` so that only the bits based on target OS will be pulled. Please see netty doc: https://netty.io/wiki/forked-tomcat-native.html#how-to-download-netty-tcnative-boringssl-static

File: sdk/core/azure-core/pom.xml [hunk 1]
```diff
@@ -88,6 +88,11 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-tcnative-boringssl-static</artifactId>
```
Comment (srnagar) [inline sdk/core/azure-core/pom.xml]: Is this verified to work with OkHttp too? I know OkHttp supports boringssl but we should verify if the versions are compatible.

**Comment by JonathanGiles** [comment]
I agree with @anuchandy - please investigate the document he links to and explore using the operating system classifier so that only the bits for a single OS are used. Weigh up the pros and cons of this approach (particularly in terms of ease of deployment for end users), and let us know your thoughts.

**Comment by JonathanGiles** [comment]
@g2vinay Can you give us a status update on this when you have further progress? Thanks!

**Comment by g2vinay** [comment]
@JonathanGiles , I'll push updates soon.

**Comment by g2vinay** [comment]
Storage Perf Numbers   (with this PR   vs GA version)
Windows VM 64bit, 2.30 GHz (WIndows 10 Pro)
Ram 7.00 GB

Processor: 
Storage Version         Parallel            API            FileSize             RPS
12.0.0                         32              download          1MB              42.3
12.1.0-beta.1             32               download          1MB             108.2     (155.79% increase from GA ver)

12.0.0                         32            uploadBlockBlob 1MB              50.7
12.1.0-beta.1             32             uploadBlockBlob 1MB             149.2     (194.28% increase from GA ver)

@JonathanGiles @joshfree @anuchandy 

@anuchandy , as disucssed we will not be using classifier due to reliability issues with maven os plugin, we wouldn't want our users to deal with.

File: sdk/core/azure-core/pom.xml (hunk 1 above)
Comment (g2vinay) [reply sdk/core/azure-core/pom.xml]: The versions are compatible,
but with okhttp one more dependency is needed and requires the user to manually setup the Security provider in the code.
Best to handle this in a future PR when okhttp performance becomes important.
and currently okhttp has some issues like resource leaks and exceptions thrown with storage SDK, Need to get it stable before focusing on its perf.
@JonathanGiles

File: parent/pom.xml [hunk 2]
```diff
@@ -194,6 +195,7 @@
     <maven-eclipse-plugin.version>2.8</maven-eclipse-plugin.version>
     <maven-assembly-plugin.version>2.2</maven-assembly-plugin.version>
     <maven-surefire-plugin.version>3.0.0-M3</maven-surefire-plugin.version>
+    <os-maven-plugin.version>1.4.0.Final</os-maven-plugin.version>
```
Comment (g2vinay) [inline parent/pom.xml]: remove this

File: sdk/core/azure-core/pom.xml [hunk 3]
```diff
@@ -96,6 +100,10 @@
       <artifactId>junit-jupiter-engine</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
```
Comment (g2vinay) [inline sdk/core/azure-core/pom.xml]: remove this, not intended

File: sdk/keyvault/azure-security-keyvault-secrets/README.md [hunk 4]
```diff
@@ -69,6 +69,28 @@ HttpClient client = new NettyAsyncHttpClientBuilder()
...
+    <exclusions>
+      <exclusion>
+        <groupId>io.netty</groupId>
+        <artifactId>netty-tcnative-boringssl-static</artifactId>
+      </exclusion>
+    </exclusions>
+</dependency>
+```
+[//]: # ({x-version-update-end})
+
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-secrets/README.md]: @JonathanGiles @anuchandy , @srnagar 
Can you review this piece, 
Let me know if I should add something more, 
Will then add it to all track 2 SDKs

File: sdk/core/azure-core/pom.xml (hunk 1 above)
Comment (JonathanGiles) [reply sdk/core/azure-core/pom.xml]: That is fine - just file a tech debt issue to track enabling SSL for OkHttp

File: sdk/keyvault/azure-security-keyvault-secrets/README.md [hunk 5]
```diff
@@ -69,6 +69,27 @@ HttpClient client = new NettyAsyncHttpClientBuilder()
     .build();
 ```
 
+### Default SSL library
+All client libraries, by default, use Tomcat native Boring SSL. The Boring SSL library is an uber jar containing bits of linux/mac/windows and provides better performance compared to default jdk SSL.
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-security-keyvault-secrets/README.md]:
```suggestion
All client libraries, by default, use the Tomcat-native Boring SSL library to enable native-level performance for SSL operations. The Boring SSL library is an uber jar containing native libraries for Linux / macOS / Windows, and provides better performance compared to the default SSL implementation within the JDK. For more information, including how to reduce the dependency size, refer to the [performance tuning](https://github.com/Azure/azure-sdk-for-java/wiki/Performance-Tuning) section of the wiki.
```

File: sdk/keyvault/azure-security-keyvault-secrets/README.md (hunk 4 above)
Comment (JonathanGiles) [reply sdk/keyvault/azure-security-keyvault-secrets/README.md]: With my text changes above, all of this section should be deleted.

Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-secrets/README.md]: done

File: sdk/core/azure-core/pom.xml (hunk 1 above)
Comment (g2vinay) [reply sdk/core/azure-core/pom.xml]: Opened the issue - #6266

**Pull request merged**